subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/codecs/ImmutableMapCodec.kt

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.instantexecution.serialization.codecs
+
+import com.google.common.collect.ImmutableMap
+import org.gradle.instantexecution.serialization.Codec
+import org.gradle.instantexecution.serialization.ReadContext
+import org.gradle.instantexecution.serialization.WriteContext
+import org.gradle.instantexecution.serialization.writeMap
+
+
+object ImmutableMapCodec : Codec<ImmutableMap<Any, Any>> {
+
+    override suspend fun WriteContext.encode(value: ImmutableMap<Any, Any>) {
+        writeMap(value)
+    }
+
+    override suspend fun ReadContext.decode(): ImmutableMap<Any, Any>? {
+        val size = readSmallInt()
+        val builder = ImmutableMap.builderWithExpectedSize<Any, Any>(size)
+        for (i in 0 until size) {
+            val key = read()
+            val value = read()
+            builder.put(key!!, value!!)
```

Comment:
⭕️ consider moving the not-null assertions to the `read()!!` calls so the exceptions, if any, carry a distinct line number in each case.

Reply:
Done!